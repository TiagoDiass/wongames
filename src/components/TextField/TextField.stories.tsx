import { ComponentMeta, ComponentStory } from '@storybook/react';
import { Email as EmailIcon } from 'styled-icons/material-outlined';
import TextField from './TextField';

export default {
  title: 'TextField',
  component: TextField,
  args: {
    label: 'Email',
    labelFor: 'email',
    initialValue: '',
    placeholder: 'jhon.doe@gmail.com'
  },
  argTypes: {
    onInput: { action: 'changed' },
    icon: { type: 'symbol' }
  }
} as ComponentMeta<typeof TextField>;

export const Basic: ComponentStory<typeof TextField> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField {...args} />
  </div>
);

export const WithIcon: ComponentStory<typeof TextField> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField {...args} autoComplete='off' />
  </div>
);

WithIcon.args = {
  icon: <EmailIcon />
};
