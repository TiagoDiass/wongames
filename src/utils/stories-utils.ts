type Backgrounds = 'Dark' | 'Light' | 'Light Blue';

export const setStoryDefaultBackground = (background: Backgrounds) => ({
  backgrounds: {
    default: background
  }
});

type Layouts = 'fullscreen' | 'centered' | 'padded';

export const setStoryLayout = (storyLayout: Layouts) => ({
  layout: storyLayout
});
