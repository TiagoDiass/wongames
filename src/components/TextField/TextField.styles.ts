import styled, { css, DefaultTheme } from 'styled-components';
import { TextFieldProps } from './TextField';

const wrapperModifiers = {
  disabled: (theme: DefaultTheme) => css`
    ${Label}, ${Input}, ${Icon} {
      color: ${theme.colors.gray};
      cursor: not-allowed;

      &::placeholder {
        color: currentColor;
      }
    }
  `
};

export const Wrapper = styled.div<Pick<TextFieldProps, 'disabled'>>`
  ${({ theme, disabled }) => css`
    ${disabled && wrapperModifiers.disabled(theme)}
  `}
`;

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.black};
    cursor: pointer;
  `}
`;

const inputWrapperModifiers = {
  iconOnRight: () => css`
    flex-direction: row-reverse;
  `
};

type InputWrapperProps = Pick<TextFieldProps, 'iconPosition'>;

export const InputWrapper = styled.div<InputWrapperProps>`
  ${({ theme, iconPosition }) => css`
    display: flex;
    background: ${theme.colors.lightGray};
    border-radius: 0.2rem;
    padding: 0 ${theme.spacings.xxsmall};
    border: 0.2rem solid;
    border-color: ${theme.colors.lightGray};

    transition: all ${theme.transition.default};

    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }

    ${iconPosition === 'right' && inputWrapperModifiers.iconOnRight()}
  `}
`;

export const Input = styled.input`
  ${({ theme }) => css`
    color: ${theme.colors.black};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall};
    background: transparent;
    border: 0;
    outline: none;
    width: 100%;
  `}
`;

export const Icon = styled.div`
  ${({ theme }) => css`
    display: flex;
    width: 2.2rem;
    color: ${theme.colors.gray};

    & > svg {
      width: 100%;
    }
  `}
`;
