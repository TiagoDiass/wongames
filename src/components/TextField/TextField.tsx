import { ChangeEvent, InputHTMLAttributes, useState } from 'react';
import * as S from './TextField.styles';

type TextFieldProps = {
  onInput?: (value: string) => void;
  label?: string;
  labelFor?: string;
  initialValue?: string;
} & InputHTMLAttributes<HTMLInputElement>;

/**
 * TextField component that is used for user inputs
 */
export default function TextField({
  label,
  labelFor = '',
  initialValue = '',
  onInput,
  ...props
}: TextFieldProps) {
  const [value, setValue] = useState(initialValue);

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    const newValue = event.target.value;

    setValue(newValue);

    !!onInput && onInput(newValue);
  };

  return (
    <S.Wrapper>
      {!!label && <S.Label htmlFor={labelFor}>{label}</S.Label>}

      <S.InputWrapper>
        <S.Input type='text' id={labelFor} value={value} onChange={handleChange} {...props} />
      </S.InputWrapper>
    </S.Wrapper>
  );
}
