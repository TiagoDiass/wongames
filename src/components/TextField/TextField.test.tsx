import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { renderWithTheme } from 'utils/test-utils';
import TextField from './TextField';

describe('<TextField />', () => {
  it('should render correctly with Label', () => {
    renderWithTheme(<TextField label='Label' labelFor='Field' id='Field' />);

    expect(screen.getByLabelText('Label')).toBeInTheDocument();
  });

  it('should render correctly without Label', () => {
    renderWithTheme(<TextField />);

    expect(screen.queryByLabelText('Label')).not.toBeInTheDocument();
  });

  it('should render correctly with placeholder', () => {
    renderWithTheme(<TextField placeholder='Type your email in here' />);

    expect(screen.getByPlaceholderText('Type your email in here')).toBeInTheDocument();
  });

  it('should call props.onInput correctly when input value changes', async () => {
    const onInput = jest.fn();
    renderWithTheme(
      <TextField onInput={onInput} label='TextField' labelFor='TextField' id='TextField' />
    );

    const input = screen.getByRole('textbox');
    const text = 'This is my new text';
    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).toHaveValue(text);
      expect(onInput).toHaveBeenCalledTimes(text.length);
    });
    expect(onInput).toHaveBeenLastCalledWith(text);
  });

  it('should be accessible with the tab key', () => {
    renderWithTheme(<TextField label='TextField' labelFor='TextField' id='TextField' />);

    const input = screen.getByLabelText('TextField');
    expect(document.body).toHaveFocus();

    userEvent.tab();
    expect(input).toHaveFocus();
  });
});
