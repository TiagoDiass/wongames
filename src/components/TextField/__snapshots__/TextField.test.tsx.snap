// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should render correctly with an accessible error 1`] = `
.c0 .c2 {
  border: 1px solid #FF6347;
}

.c0 .c2:focus-within {
  border: 2px solid #FF6347;
  box-shadow: none;
}

.c1 {
  font-size: 1.4rem;
  color: #030517;
  cursor: pointer;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #EAEAEA;
  border-radius: 0.2rem;
  padding: 0 0.8rem;
  border: 0.2rem solid;
  border-color: #EAEAEA;
  -webkit-transition: box-shadow 0.3s ease-in-out;
  transition: box-shadow 0.3s ease-in-out;
}

.c3:focus-within {
  border: 1px solid #F231A5;
}

.c4 {
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c5 {
  color: #FF6347;
  font-size: 1.2rem;
}

<div
  class="c0"
>
  <label
    class="c1"
    for="email"
  >
    Label
  </label>
  <div
    class="c2 c3"
  >
    <input
      aria-disabled="false"
      aria-errormessage="email-error"
      aria-invalid="true"
      class="c4"
      id="email"
      type="text"
      value=""
    />
  </div>
  <span
    class="c5"
    id="email-error"
  >
    Please fill out this field
  </span>
</div>
`;
